docs/digits_vae.md

```diff
@@ -0,0 +1,513 @@
+    self.decoder = Decoder(latent_size, hidden_size, input_size, rngs=rngs)
+
+  def __call__(self, x: jax.Array) -> tuple[jax.Array, jax.Array, jax.Array]:
+    x = jax.vmap(jax.numpy.ravel)(x)  # flatten
+    z, mean, std = self.encoder(x)
+    logits = self.decoder(z)
+    logits = jnp.reshape(logits, (-1, *self.image_shape))
+    return logits, mean, std
+```
+
++++ {"id": "xIm9Yi5YoIxN"}
+
+Next is the loss function – there are two components to the model that we want to ensure:
+
+1. the `logits` output faithfully reconstruct the input image.
+2. the model represented by `mean` and `std` faithfully represents the "true" latent distribution.
+
+VAE uses a loss function based on the [Evidence lower bound](https://en.wikipedia.org/wiki/Evidence_lower_bound) to quantify theset two goals in a single loss value:
+
+```{code-cell}
+:id: bMpxj8-Wsvui
+
+def vae_loss(model: VAE, x: jax.Array):
+  logits, mean, std = model(x)
+  kl_loss = jnp.mean(0.5 * jnp.mean(
+      -jnp.log(std ** 2) - 1.0 + std ** 2 + mean ** 2, axis=-1))
+  reconstruction_loss = jnp.mean(
+    optax.sigmoid_binary_cross_entropy(logits, x)
+  )
+  return reconstruction_loss + 0.1 * kl_loss
+```
+
++++ {"id": "RaT0ELpvqo2W"}
+
+Now all that's left is to define the model and optimizer, and run the training loop:
+
+```{code-cell}
+:id: JPgoHL5rpKXd
+:outputId: e7626646-7c14-42dc-c18f-774adff1306e
+
+import optax
+
+model = VAE(
+  image_shape=(8, 8),
+  hidden_size=32,
+  latent_size=8,
+  rngs=nnx.Rngs(0, noise=1),
+)
+
+optimizer = nnx.Optimizer(model, optax.adam(1e-3))
+
+@nnx.jit
+def train_step(model: VAE, optimizer: nnx.Optimizer, x: jax.Array):
+  loss, grads = nnx.value_and_grad(vae_loss)(model, x)
+  optimizer.update(grads)
+  return loss
+
+for epoch in range(2001):
+  loss = train_step(model, optimizer, images_train)
+  if epoch % 500 == 0:
+    print(f'Epoch {epoch} loss: {loss}')
+```
+
++++ {"id": "f8m_uoL9q47M"}
+
+And here we see that something has gone wrong: our loss value has become NaN after some number of iterations.
+
++++ {"id": "SBS1mmxwrS25"}
+
+## Debugging NaNs in JAX
+Despite our best efforts, our model is producing NaNs. What can we do?
+
+JAX offers a number of debugging approaches for situations like this, outlined in the JAX docs at [Debugging runtime values](https://jax.readthedocs.io/en/latest/debugging/index.html). In this case we can use the [`debug_nans`](https://jax.readthedocs.io/en/latest/debugging/flags.html#jax-debug-nans-configuration-option-and-context-manager) configuration to see where the NaN value is arising:
+
+```{code-cell}
+:id: JE7OYoZ4rRQ8
+:outputId: 2b77a9c1-aaa4-418a-da71-013d8f885340
+
+model = VAE(
+  image_shape=(8, 8),
+  hidden_size=32,
+  latent_size=8,
+  rngs=nnx.Rngs(0, noise=1),
+)
+
+optimizer = nnx.Optimizer(model, optax.adam(1e-3))
+
+with jax.debug_nans(True):
+  for epoch in range(2001):
+    train_step(model, optimizer, images_train)
+```
```

Comment:
`Invalid nan value encountered in the output of a C++-jit/pmap function. Calling the de-optimized version.`

Not for this PR, I feel like we can improve this message (e.g. why does it mention C++? And we have shard_map now)

Reply:
Also I wonder if we can somehow make the stack trace shorter... I'm guessing probably not, but why are we reraising the exception so many times?

Also `invalid value (nan) encountered in pjit` needs to be updated

Reply:
Yeah agreed this error is really hard to parse. Maybe some of it could be fixed by using `tracebackutil.register_exclusion` within parts of NNX

Reply:
cc/ @cgarciae – can we take a look at how tracebacks interact with NNX transformations?